Mathlib/CategoryTheory/Functor/Flat.lean

```diff
@@ -0,0 +1,597 @@
+/-
+Copyright (c) 2021 Andrew Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Andrew Yang
+
+! This file was ported from Lean 3 source module category_theory.functor.flat
+! leanprover-community/mathlib commit 14e80e85cbca5872a329fbfd3d1f3fd64e306934
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.CategoryTheory.Limits.FilteredColimitCommutesFiniteLimit
+import Mathlib.CategoryTheory.Limits.Preserves.FunctorCategory
+import Mathlib.CategoryTheory.Limits.Bicones
+import Mathlib.CategoryTheory.Limits.Comma
+import Mathlib.CategoryTheory.Limits.Preserves.Finite
+import Mathlib.CategoryTheory.Limits.Shapes.FiniteLimits
+/-!
+# Representably flat functors
+
+We define representably flat functors as functors such that the category of structured arrows
+over `X` is cofiltered for each `X`. This concept is also known as flat functors as in [Elephant]
+Remark C2.3.7, and this name is suggested by Mike Shulman in
+https://golem.ph.utexas.edu/category/2011/06/flat_functors_and_morphisms_of.html to avoid
+confusion with other notions of flatness.
+
+This definition is equivalent to left exact functors (functors that preserves finite limits) when
+`C` has all finite limits.
+
+## Main results
+
+* `flat_of_preserves_finite_limits`: If `F : C ⥤ D` preserves finite limits and `C` has all finite
+  limits, then `F` is flat.
+* `preserves_finite_limits_of_flat`: If `F : C ⥤ D` is flat, then it preserves all finite limits.
+* `preserves_finite_limits_iff_flat`: If `C` has all finite limits,
+  then `F` is flat iff `F` is left_exact.
+* `Lan_preserves_finite_limits_of_flat`: If `F : C ⥤ D` is a flat functor between small categories,
+  then the functor `Lan F.op` between presheaves of sets preserves all finite limits.
+* `flat_iff_Lan_flat`: If `C`, `D` are small and `C` has all finite limits, then `F` is flat iff
+  `Lan F.op : (Cᵒᵖ ⥤ Type*) ⥤ (Dᵒᵖ ⥤ Type*)` is flat.
+* `preserves_finite_limits_iff_Lan_preserves_finite_limits`: If `C`, `D` are small and `C` has all
+  finite limits, then `F` preserves finite limits iff `Lan F.op : (Cᵒᵖ ⥤ Type*) ⥤ (Dᵒᵖ ⥤ Type*)`
+  does.
+
+-/
+
+
+universe w v₁ v₂ v₃ u₁ u₂ u₃
+
+open CategoryTheory
+
+open CategoryTheory.Limits
+
+open Opposite
+
+namespace CategoryTheory
+
+namespace StructuredArrowCone
+
+open StructuredArrow
+
+variable {C : Type u₁} [Category.{v₁} C] {D : Type u₂} [Category.{v₁} D]
+
+variable {J : Type w} [SmallCategory J]
+
+variable {K : J ⥤ C} (F : C ⥤ D) (c : Cone K)
+
+-- **TODO** Scott changed `@[simps]` to `@[simps!]` below and I don't
+-- know what this does, but one thing it does is that
+-- it stops `toDiagram_obj` being created, and `toDiagram_obj` is
+-- used later on so I (kmb) have changed it back
```

Comment:
@semorrison  In [this commit](https://github.com/leanprover-community/mathlib4/pull/3703/commits/212f7539353e0a8327b292354cfd2676488f2038) in the port of CategoryTheory.Functor.Flat you changed the `@[simps]` on `toDiagram` to `@[simps!]`. This breaks the proof of `uniq` (see line 249 or so). I don't know what `@[simps!]` does (or how to find out) but I thought I'd flag this.

Reply:
Thanks, `simps!` was wrong. It's for cases where `simps` needs to unfold a definition before it finds a constructor (like `{rhs_md := reducible}` in mathlib3). Here it's just a constructor, sorry I don't know why I changed it. 🤷‍♀️